### PR TITLE
Recipe to add a new language to Llama

### PR DESCRIPTION
**Pull request by rahul-sarvam**
This PR adds a recipe to extend Llama tokenizer with tokens from a new language and use the extended tokenizer to continue pretraining / finetuning. The major changes are as follows:
- added a new folder inside recipes called `multilingual` with code required to extend the tokenizer
- added a new train config parameter called `tokenizer_name` in `src/llama_recipes/configs/training.py` so that a custom tokenizer can be used in `recipes/finetuning/finetuning.py`
- added code to `recipes/finetuning/finetuning.py` to resize the model's embeddings if the tokenizer size is greater than the size of the model's embedding matrix

### CODE REVIEW

File: recipes/multilingual/README.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# Extending Llama to a new language
+
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:2]: @rahul-sarvam shall we add a sentences that this work and dataset has been done by Saravam, and perhaps we can link the blog post as well.

File: recipes/multilingual/README.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+In this recipe, we will see how to add a new language to the Llama family of models. The steps are quite general and can be easily adapted to other models as well. By default, we will use the [Varta](https://huggingface.co/datasets/rahular/varta) dataset and add Hindi to Llama2. You can add any other language present in the dataset by only passing the right language code (advanced users can also tweak the code to add multiple languages at once).
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:3]: It will be great to add bit of explanation about this dataset, linking any blog post about it. Also some of the story if special cleaning/ preprocessing steps has been done to create this dataset ( adding some of the useful tooling) that help the community to further understand their options/ steps they might need to think building a dataset.

File: recipes/multilingual/README.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+In this recipe, we will see how to add a new language to the Llama family of models. The steps are quite general and can be easily adapted to other models as well. By default, we will use the [Varta](https://huggingface.co/datasets/rahular/varta) dataset and add Hindi to Llama2. You can add any other language present in the dataset by only passing the right language code (advanced users can also tweak the code to add multiple languages at once). 
+
+The first thing we will do is create a new tokenizer whose vocabulary contains only tokens from Hindi. For this, we will first download and prepare the data for training the tokenizer:
+
+```
+python prepare_data.py --split=validation --lang=hi --docs_to_sample=10000 --save_path=./data
+```
+
+Here we sample 10,000 Hindi documents from the validation split (we should ideally sample from the training split, but this is much faster) and save it as a text file inside `./data`. Next, we use this text to train a Hindi-only [sentencepiece](https://github.com/google/sentencepiece) tokenizer with a vocabulary size of 16,000.
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:11]: maybe we could also some of the ideas around how to select vocab size for tokenizer and if any specific setting in sentence piece etc, maybe something similar to [this](https://github.com/HamidShojanazeri/notes/blob/main/tokenizer_vocab_size.md)based on your experience.

File: recipes/multilingual/extend_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import fire
+import re
+from transformers import LlamaTokenizer
+
+os.environ["PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION"] = "python"
+from huggingface_hub import hf_hub_download
+from sentencepiece import sentencepiece_model_pb2 as sp_pb2_model
+
+
+def main(new_tokenizer_path, extended_tokenizer_save_path):
```
Comment (HamidShojanazeri) [inline recipes/multilingual/extend_tokenizer.py:15]: it would be great if we could add some comment in the code for different step/ nuances.

File: recipes/multilingual/README.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+...     assert llama_tokenizer.convert_ids_to_tokens(i) == our_tokenizer.convert_ids_to_tokens(i), f"Token mismatch at index {i}."
+...
+>>> text = "मैं एक अच्छा हाथी हूँ"
+>>> llama_tokenizer.tokenize(text)
+['▁', 'म', 'ै', 'ं', '▁', '<0xE0>', '<0xA4>', '<0x8F>', 'क', '▁', 'अ', 'च', '्', '<0xE0>', '<0xA4>', '<0x9B>', 'ा', '▁', 'ह', 'ा', 'थ', 'ी', '▁', 'ह', 'ू', '<0xE0>', '<0xA4>', '<0x81>']
+>>> our_tokenizer.tokenize(text)
+['▁मैं', '▁एक', '▁अच', '्', 'छा', '▁हाथी', '▁हूँ']
+```
+
+Finally, we can use this tokenizer to finetune Llama2 on Hindi datasets by following the (finetuning recipes)[https://github.com/subramen/llama-recipes/tree/new-folder-structure/recipes/finetuning]. Remember to pass the new tokenizer path as an argument to the script: `--tokenizer_name=./extended_tokenizer`.
```
Comment (HamidShojanazeri) [inline recipes/multilingual/README.md:43]: It would be great if we could also add, some more details, how much we need to train it, any hyper parameter settings etc. I would be more than happy to run the experiments and add a bunch of loss curves etc to make more comprehensive as an e2e recipe.

**Review by HamidShojanazeri — COMMENTED** [review]
Thanks so much @rahul-sarvam for the great PR!! added some minor comments and will be more than happy to help incorporating any of those points with your guidance.

File: recipes/multilingual/README.md (hunk 1 above)
Comment (amitsangani) [reply recipes/multilingual/README.md:2]: Yes, I agree that we should attribute to Sarvam here!

File: recipes/multilingual/README.md (hunk 5 above)
Comment (HamidShojanazeri) [reply recipes/multilingual/README.md:43]: @rahul-sarvam I also wonder if we need to reflect the vocab size change in the model embedding layer/ init the added ones with random weights, shall we have some guidance on that part too.
